src/NetTopologySuite/Algorithm/RobustLineIntersector.cs

```diff
@@ -326,119 +316,27 @@ private static Coordinate NearestEndpoint(Coordinate p1, Coordinate p2,
             return nearestPt;
         }
 
-        /*
-         * Gets the Z value of the first argument if present, 
-         * otherwise the value of the second argument.
-         * 
-         * @param p a coordinate, possibly with Z
-         * @param q a coordinate, possibly with Z
-         * @return the Z value if present
-         */
-        private static double zGet(Coordinate p, Coordinate q)
-        {
-            double z = p.Z;
-            if (double.IsNaN(z))
-            {
-                z = q.Z; // may be NaN
-            }
-            return z;
-        }
+        public ElevationModel ElevationModel { get; set; }
 
         /// <summary>
-        /// Gets the Z value of a coordinate if present, or
-        /// interpolates it from the segment it lies on.
-        /// If the segment Z values are not fully populate
-        /// NaN is returned.
+        /// Gets or sets a value indicating if a local <see cref="Algorithm.ElevationModel"/> should be computed from
+        /// the coordinates involved in the robust line intersection test, if other means to
+        /// get one fail. 
         /// </summary>
-        /// <param name="p">A coordinate, possibly with Z</param>
-        /// <param name="p1">A segment endpoint, possibly with Z</param>
-        /// <param name="p2">A segment endpoint, possibly with Z</param>
-        /// <returns>The extracted or interpolated Z value (may be NaN)</returns>
-        private static double zGetOrInterpolate(Coordinate p, Coordinate p1, Coordinate p2)
-        {
-            double z = p.Z;
-            if (!double.IsNaN(z))
-                return z;
-            return zInterpolate(p, p1, p2); // may be NaN
-        }
+        public static bool UseLocalElevationModel { get; set; } = true;
```

Comment:
I see now what you were saying in the description. Yeah, using global `static` variables like this to alter the behavior of some specific methods is quite a bit limiting. *I actually personally hate it tremendously, but I do recognize that it can be situationally useful if there's a client with **very** specific needs that shouldn't make all other API consumers have to suffer for it.*

A few thoughts on alternatives:
- Putting an `ElevationModel` onto `NtsGeometryServices` is hypothetically useful for more than just this, the `ElevationModel` API doesn't look horrible (especially with the suggested changes), and "`null` means to use a local one that's specific to the current operation" gives the same amount of information as a `bool` flag while **also** making it easier for people to swap in their own implementation if they have different ideas.
- Even if it doesn't seem viable or a good idea to put a whole `ElevationModel` onto `NtsGeometryServices`, we should **at least** be able to put the `bool` flag there so that it's more discoverable (unless we really **don't want** it to be more discoverable, and that's part of the point?).
- In **some** places where we create a `RobustLineIntersector` (not most, it seems), we have very easy access to a `Geometry` whose `g.Factory.GeometryServices` is (or should be) allowed to have a different opinion on what to use than some potentially unrelated calls in the same application, so in those cases the user wouldn't need to copy-paste a ton of code from our repository just to get a specific higher-order operation to run using a different `ElevationModel`.
- I imagine that in a lot of the cases where we don't have a `Geometry`, there still **was** a `Geometry` somewhere in the chain of operations that led to it, and its `g.Factory.GeometryServices` could get passed down through some new overloads along that chain to make that viable for even more cases.
- Of course, if someone wants to call `LineSegment.Intersection(LineSegment)`, then we will have to make **some** decision, and I don't want that decision to be "the API was deleted", so there's no escaping magical static variables like `NtsGeometryServices.Instance` sometimes.